### DISCUSSION TITLE
Should I manually free char* pointers returned by magick wand api methods?

### DISCUSSION THREAD

**Discussion by Haidar0096** [ImageMagick/ImageMagick#5706] [Development] [answered]
The [documentation](https://imagemagick.org/api/magick-wand.php) is not helpful at all in this regards.

 For example, [MagickGetException](https://imagemagick.org/api/MagickWand/magick-wand_8c.html) returns a `char*`. So what should I do with it when I am done? Should I free it or will the api free it when I destroy the wand?

**Comment by dlemstra** [comment, marked as answer]
Yes you will need to free that yourself. https://github.com/ImageMagick/ImageMagick/blob/main/MagickWand/magick-wand.c#L282. When we return a `char *` it becomes your responsibility to free it. This is not the case for `const char *`.

**Reply by Haidar0096** [reply]
> Yes you will need to free that yourself. https://github.com/ImageMagick/ImageMagick/blob/main/MagickWand/magick-wand.c#L282. When we return a `char *` it becomes your responsibility to free it. This is not the case for `const char *`.

@dlemstra  I saw the code you referred before, but I couldn't keep going deeper into the core library code to see if you are freeing it somewhere else or not.

Anyway, does this apply to all `x*` (`x` can be any type) returned from other functions of the api? for example, [DestroyMagickWand](https://imagemagick.org/api/MagickWand/magick-wand_8c.html) is documented as "DestroyMagickWand() deallocates memory associated with an MagickWand.", but for some reason unknown to me, it returns some pointer to a `MagickWand`, now what is this pointer, and should I free it or not?

pardon me but I am worried to have memory leaks in my app if I don't free something I am supposed to free.

**Reply by dlemstra** [reply]
The short answer to your question is yes. When `x*` is return you will need to free it. But the exception in this case is `DestroyMagickWand` because it will return a null value that you can use to set your wand value to null instead of uninitialized memory. After this call `wand=DestroyMagickWand(wand)` the value of `wand` becomes `NULL`. All the `DestroyXX` methods that we have use this pattern.

**Reply by Haidar0096** [reply]
okay thank you for your help :)
